doc/tutorials/ostro-oic-tutorial.md

```diff
@@ -0,0 +1,1030 @@
+```
+
+OIC resources, like Soletta light-server sample, have an ID to uniquely
+identify them. As our client is going to connect to a specific resource,
+we need to define its ID somehow. We are going to get them from command
+line argument used to run our application. This can be done with help
+of node `app/argc-argv`:
+
+```
+cmdline_arguments(app/argc-argv)
+_(constant/int:value=1) OUT -> IN cmdline_arguments
+```
+
+With this, when we get output value from `cmdline_arguments` node, it
+will output command line argument number 1.
+Then, we use Soletta `oic/client-light` [5] to communicate with a light
+server. We also cross link it our controller, so changes on one are
+properly reflected on another - note however, that at this point, changes
+only happen in the direction 'controller -> oic/client', but we'll
+improve this later.
+
+```
+oic(oic/client-light)
+
+# Uses command line argument as device id
+cmdline_arguments OUT -> DEVICE_ID oic
+
+# Cross link OIC and controller
+oic OUT_STATE -> STATE controller
+oic OUT_POWER -> POWER controller
+
+controller STATE -> IN_STATE oic
+controller POWER -> IN_POWER oic
+```
+
+Let's not forget to show user about server status:
+
+```
+oic OUT_STATE -> IN _(console:prefix="Server state is: ")
+oic OUT_POWER -> IN _(console:prefix="Server power level is: ")
+```
+
+### Showing a nice error message to user
+
+If we run our `controller-oic.fbp` as is, it will end up showing an error
+message like:
+
+```
+$ sol-fbp-runner controller-oic.fbp
+
+WRN:sol-flow ./src/lib/flow/sol-flow-static.c:341 flow_send_do() Error packet '22 (Argument position (1) is greater than arguments length (1))' sent from 'controller-oic.fbp (0x560d1c548e90)' was not handled
+```
+
+This is so because we didn't pass the device id as argument. But how a
+user would know? Every Soletta node type as an implicit `ERROR` port,
+where error packets are sent. The idea will be capture an ERROR packet
+sent by `cmdline_arguments` and then print a more meaningful message to
+user:
+
+```
+cmdline_arguments ERROR -> IN cmdline_usage(converter/empty-to-string:output_value="Missing device ID in the command line. Example: 2f3089a1dbfb43d38cab64383bdf9380") OUT -> IN _(console:prefix="ERROR: ")
+cmdline_usage OUT -> IN _(converter/empty-to-int:output_value=1) OUT -> CODE _(app/quit)
+```
+
+See that we used an `app/quit` node type to quit application with an
+error code, in this case, 1.
+
+Now, if we run without the expected argument, we get a better error
+message:
+
+```
+$ sol-fbp-runner controller-oic.fbp
+
+ERROR: Missing device ID in the command line. Example: 2f3089a1dbfb43d38cab64383bdf9380 (string)
+```
+
+### Running the server
+
+To really test our OIC controller, we need to run Soletta
+`light-server.fbp` sample, using the device id we are going to pass
+to the application. To do so, you can run, on Soletta dir:
+
+```
+SOL_MACHINE_ID=2f3089a1dbfb43d38cab64383bdf9380 sol-fbp-runner src/samples/flow/oic/light-server.fbp
```

Comment:
not good to require the `SOL_MACHINE_ID`, we'd have to explain that detail. it's better to immediately change `light-server` to print it and then ask user to use that value.